### ISSUE TITLE
While users are logged off, tileEntitiesToBeRemoved grows to large size

### ISSUE THREAD

**Issue by Ricket** [MinecraftForge/MinecraftForge#4238]
Minecraft 1.11.2, forge version 13.20.1.2393 (modpack All the Mods 2 v0.73)

I'm having an issue on my server, where occasionally any user is unable to log in (they get timed out) and then the user appears to still be logged in for some minutes after, and then eventually the user is logged out and they can log in normally.

I was able to catch it while profiling, so I found these removeAll methods took a really, really long time and are blocking the server thread:

![image](https://user-images.githubusercontent.com/196978/28638756-d2cd70f8-7213-11e7-9f75-2a78565a460e.png)

This of course results in the timeout, and the server is basically stuck until removeAll completes.

I found this corresponds to World.updateEntities, specifically these two lines:

```
this.tickableTileEntities.removeAll(this.tileEntitiesToBeRemoved);
this.loadedTileEntityList.removeAll(this.tileEntitiesToBeRemoved);
```

I went back and looked at a memory dump I took a couple days ago, and found this:

![image](https://user-images.githubusercontent.com/196978/28643624-6d9bb832-7224-11e7-922f-4ab9051b0d35.png)

Look at those large sizes! These fields (which exist in World, the superclass of WorldServer) are:
field_147482_g = loadedTileEntityList (A list of the loaded tile entities in the world)
field_147483_b = tileEntitiesToBeRemoved
field_175730_i = tickableTileEntities

Here are the contents of those lists, at least the highest numbers of objects.

Loaded tile entity list:
![image](https://user-images.githubusercontent.com/196978/28644042-ecd7692e-7225-11e7-9495-04f185cae281.png)
![image](https://user-images.githubusercontent.com/196978/28643890-4f1619b0-7225-11e7-8d3d-643c6621273b.png)

Here is the entities to be removed:

![image](https://user-images.githubusercontent.com/196978/28643796-087b4eee-7225-11e7-9f1e-7e427aa8e3df.png)

and the tickableTileEntities:

![image](https://user-images.githubusercontent.com/196978/28643813-17ce5198-7225-11e7-8cea-fab64980d3dd.png)

There is only one ThermalEvaporationController in the world, only 9 TileEntityCokeOven, and so on. Subtract the loaded or tickable objects count from the removed entities count and you can see the real number of objects that exist.

My wild guess is that some mod is loading & unloading the chunk(s), causing those tile entities to be loaded but then added to the removed list, only to be loaded again.

What can I do about it? Is there some log where I can determine which mod is doing it?

I notice that there are multiple different mods' entities in those lists so that's why I suspect this is a forge issue. Seems like whatever loads the tile entity, should first clean up any previously-loaded entry of the same entity.

Also, why aren't the entities just periodically removed, why do they all wait and build up until a user logs in? I know this particular code is within Minecraft code, but Forge would be the one doing the chunk loading, right? (maybe not? I'm not a modder)

As a sidenote, removeAll would be a lot faster if these were HashSets instead of ArrayLists. I know Forge can patch MC source but I'm not sure of the side effects of that. Just musing that maybe if those were Sets, the removeAll might take a lot less time.

**Comment by Niko-sk2x** [comment]
This looks like a vanilla issue. `WorldServer#updateEntities` returns early without doing anything some ticks after there has been no player on the server, but world ticking is still done, which could keep loading and unloading chunks (and it does). And the tile entities are then added to the list and not cleaned up until someone joins the server.

As for the performance of it - there are already attempts to improve it (https://github.com/MinecraftForge/MinecraftForge/pull/4236)

**Comment by Ricket** [comment]
Is it an issue that chunks seem to be loaded and unloaded repeatedly? Is there a good way to debug what entity/mod is causing the toggling loads/unloads?

Could Forge patch WorldServer to remove that idle behavior? I guess it's trying to reduce CPU usage, but I'd much rather keep ticking 24/7 if it solves this issue. Seems like the check in WorldServer.updateEntities could just be removed.

Thanks for pointing out the pull request, seems he took it even a step further than my suggestion. That might make the issue "go away" from the standpoint of the login timeouts (because the list cleanup will happen quick enough), but there is still the issue that memory usage will keep growing while the server is not ticking but chunks are loading/unloading. I think our server would be ok for now, but if there were more entities in those chunks then it could definitely become an issue.

**Comment by asiekierka** [comment]
@Ricket If you need a quick workaround, my initial analysis to the code seems to imply that placing a single chunkloader working constantly anywhere in the world should do it.

**Comment by Ricket** [comment]
Thanks, yeah I chunkloaded the affected chunk area last night (though I now see that you're right it could have been placed anywhere). It did solve the problem this morning. But then I took a memory dump just now and the issue was back! After some more investigation I realized that the chunkloader I used (A Chunk Loader) has a time period, and in my pack that time was set to 8 hours. So once the 8 hours expired, the chunkloader released its Ticket and the problem resumed. I reconfigured it to be unlimited time and I am going to move the chunkloader to a random chunk.

I am still curious about my question of what mod/block is doing the chunkloading. Because technically there is still going to be the issue of those chunks loading/unloading repeatedly, it's just that now the unloaded entities will be cleaned up immediately each time instead of piling up.
